bot/bot.go

```diff
@@ -47,49 +51,62 @@ func New(cfg *config.Config, dal *data.DAL, gh *github.API, log *log.Entry) *Bot
 
 	b := &Bot{
 		token:    cfg.SlackToken,
-		api:      api,
+		API:      api,
 		rtm:      api.NewRTM(),
-		dal:      dal,
-		gh:       gh,
-		log:      log,
-		commands: map[string]*cmd.Command{},
+		DAL:      dal,
+		GitHub:   gh,
+		Log:      log,
+		Commands: map[string]*cmd.Command{},
+		handlers: map[string][]EventHandler{},
 	}
 	b.PopulateUsers()
 
-	// Attach command handlers
-	b.commands = map[string]*cmd.Command{
-		"help":         NewHelpCmd(b.help),
-		"set":          NewSetCmd(b.set),
-		"edit":         NewEditUserCmd(b.editUser),
-		"view-user":    NewViewUserCmd(b.viewUser),
-		"view-team":    NewViewTeamCmd(b.viewTeam),
-		"add-user":     NewAddUserCmd(b.addUser),
-		"add-team":     NewAddTeamCmd(b.addTeam),
-		"add-admin":    NewAddAdminCmd(b.addAdmin),
-		"remove-admin": NewRemoveAdminCmd(b.removeAdmin),
-		"remove-user":  NewRemoveUserCmd(b.removeUser),
-		"remove-team":  NewRemoveTeamCmd(b.removeTeam),
-		"teams":        NewTeamsCmd(b.listTeams),
-		"refresh":      NewRefreshCmd(b.refresh),
-	}
+	// Register default Slack event handlers
+	b.RegisterEventHandlers(map[string]EventHandler{
+		"message":     b.handleMessageEvent,
+		"team_join":   b.handleUserChange,
+		"user_change": b.handleUserChange,
+	})
 	return b
 }
 
+// RegisterEventHandlers registers a handlers for different events. These
+// handlers will be called when an event of the corresponding type is received.
+func (b *Bot) RegisterEventHandlers(handlers map[string]EventHandler) {
+	for evt, handler := range handlers {
+		if b.handlers[evt] == nil {
+			b.handlers[evt] = []EventHandler{handler}
+		} else {
+			b.handlers[evt] = append(b.handlers[evt], handler)
+		}
+		b.Log.Infof("registered handler for %s Slack event", evt)
+	}
+}
+
+// RegisterCommands registers commands that the bot should handle.
+func (b *Bot) RegisterCommands(commands []*cmd.Command) {
+	for _, c := range commands {
+		if b.Commands[c.Name] != nil {
+			b.Log.Errorf("attempt to register duplicate commands: %s", c.Name)
```

Comment:
It might be best to have this fail completely and exit, and set up Travis to run `RegisterPlugins` - because a registered plugins can change independently of each other as part of different repos, it might be good to automate this check and warn of it more explicitly rather than let if fail silently

Reply:
Good point! Fixed!